.github/workflows/external-code-affected.yml

```diff
@@ -0,0 +1,179 @@
+# Check if code checked into external resources (blogs, tutorials)
+# that we also track in our CI is affected by a PR.
+# In that case, we add a label to the PR (`external-code-affected`) and
+# add a comment to make sure that the external code still works and is
+# eventually updated.
+name: External code check
+
+on: pull_request_target
+
+jobs:
+  check-changes:
+    permissions: write-all
+    runs-on: ubuntu-latest
+    steps:
+    - name: Checkout repository
+      uses: actions/checkout@v2
+      with:
+        ref: ${{ github.event.pull_request.head.sha }}
+        fetch-depth: 0
+      
+    - name: Check for changes in tracked files
+      run: |
+        set -xe
```

Comment:
Is it possible to put this into a script with some test cases?

Reply:
The main problem with this is that GitHub actions protects workflows, but not script files from being tampered with.

We have to run the workflow as `pull_request_target`, which means that the (ephemeral) github token will have read/write access to the github API.

If a PR from a forked repository tries to change this workflow, it will not run - it will just run the version on the master branch instead. That's a feature of GitHub's security for workflows.

However, if we just ran a script inside the repository, this protection would not apply. The user can just change the script and execute arbitrary code with read/write access to the github API.

As an alternative, we can copy the script to somewhere else and test it there - we then just have to keep the two locations in sync.

I've did quite a bit of manual testing on the script, so right now it should be fine, but I'll leave it up to you to decide if we need more tests before landing it.

Reply:
reference link:
https://docs.github.com/en/actions/using-workflows/events-that-trigger-workflows#pull_request_target

Reply:
> in the context of the base of the pull request

this would mean that we cannot patch vulnerability?

and the script now feels quite a bit complicated.

can we tell it to download a script/program from a source that we control and just run that?

or maybe just build our own github app..

Reply:
We can develop in branches at the main Ray repo, which will also trigger the workflow on PR. The context protection only affects PRs from forked repositories.


E.g. create a branch `workflow-dev` and then file a PR for that branch. That's also what I do in this very PR btw - it's a branch on the main repository (`ci/track-external-code`) and [this PR](https://github.com/ray-project/ray/pull/35263) is a PR targeting that branch (that will be auto-rebased onto master once this PR is merged).